RavenDB-21399 Fixing lookup merge/split issues with large number of records & updates

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21399

### Additional description

We had a few issues WRT merging and splitting of branch pages when the number of records exceeded 1.7M and 3.4M items.
Updating specific items in the tree (leftmost item in a branch page) would merge improperly and sometimes lead to a missing value.

Reworked the way this is handled, added additional diagnostics and made the process simpler.

### Tests

Testes were added to ensure that the behavior is valid.

ayende committed:
RavenDB-21321 Making sure that we update the state variable after we move to the next page, if the _depth_ of the tree isn't the same, we may look at the wrong value

ayende committed:
RavenDB-21399 Fixing lookup merge / split processes.
- Allow to skip leaf entries on Render to allow debugging very large trees
- Simplifying page merges, we'll only allow that when we can remove a full page, not just shuffle entries around
- During page merge, we'll try the merge on a temporary buffer, not the actual page
- Properly ensure that for branch pages, the leftmost entry is always the smallest value
- Added additional diagnostics code

ayende committed:
RavenDB-21399
- Fixed deleting a page that has no entries in lookups
- Will defrag a page always when about to do a merge between pages (already checked that the free space is big enough, so should work)
- Fixed using the wrong page as base when splitting a branch page
- Added more checks to detect problems early

arekpalinski commented:
@ayende there are some compact tree tests failures:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/8391/

ayende commented:
Having actual tests fail is awesome, managed to fix them all

ayende commented:
I run the steps on https://issues.hibernatingrhinos.com/issue/RavenDB-21399 , and there are no indexing errors.
This looks good

redknightlois commented on src/Voron/Data/Lookups/Lookup.cs:
```diff
@@ -371,180 +383,218 @@ private bool RemoveFromPage(ref TLookupKey key, bool allowRecurse)
+            // no point in trying if after combining *both* pages, we'll be completely full
+            // we also want to avoid merge / split cycles - so a merge should happen only when
+            // there is sufficient free space _after_ the merge  
+            return false;
```
I like this.

redknightlois commented:
LGTM

arekpalinski commented:
There is one failure left:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/8407/testReport/junit/SlowTests.Corax.Bugs/CompactTreeAddAndRemove/gz__/

arekpalinski commented:
test this in debug please

ayende commented:
Looking into this

ayende committed:
RavenDB-21399
- Fixing splitting of branch page in the middle not using the right position
- Adding additional verification code

ayende commented:
Fixed, and the error here is _very_ specific, so awesome that we found that in the test, but it won't have impact elsewhere
